settings: Unite user settings into a single panel.

shubham-padia opened the pull request:
Previously, there were three different sections for managing active users, deactivated users and invitations.

Fixes: #26949.

This PR builds upon #28359 and has @shashank-23002 as its co-author. Compared to the original PR, there is a bit of reorganisation of where each piece of code lives. I've also taken the liberty to not use the term `right_section` from the original PR since it was getting me confused with other `right_` components. I've renamed it to `user_settings_tab`.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
<img width="1298" alt="Screenshot 2024-05-31 at 12 53 10 PM" src="https://github.com/zulip/zulip/assets/13910561/49c90f03-a787-44c3-ac54-25d5805485ae">
<img width="1298" alt="Screenshot 2024-05-31 at 12 53 04 PM" src="https://github.com/zulip/zulip/assets/13910561/b54d7bfc-267a-4a16-bf9b-d4aa146456ca">

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

shubham-padia commented on web/src/settings_panel_menu.js:
```diff
@@ -64,22 +66,39 @@ export class SettingsPanelMenu {
+                $(".user-settings-section").hide();
+                $(`[data-user-settings-section="${CSS.escape(key)}"]`).show();
+            },
+        });
```
Initially, I had this defined in the constructor. But this lead to the toggle cliick listeners not being initialized when trying to access the settings through the gear menu from the zulip home page.

sahil839 replied:
I guess it would be better to just have this code in `settings_users.set_up_humans` as this is specifically related to the users list. And by keeping this here, this code is executed everytime the user toggles between `Personal` and `Organization` tabs in the left, which is not necessary.

shubham-padia replied:
That was how it was in the original PR, yes. But `settings_panel_menu` imports `settings_sections` which imports `settings_users`. But the toggle's callback requires `settings_panel_menu` in `settings_users` causing a circular import.

`settings_panel_menu` felt like the next best place to put it while avoiding the circular import issue.

shubham-padia replied:
Figured it out, introduced 98539623aa7b8828723ddc99eca188505cd463a6 to expose `register_event_handlers` method for the toggle component. I've moved the toggle declaration to the constructor constructor and the events are reregistered only in the case when the `section` is `user-list-admin`.

This should be ready for another review.

sahil839 replied:
I am seeing a bug where the tabs are not visible when I open the settings UI from the gear menu first time after a reload.

shubham-padia replied:
Thanks, good catch. Fixed [here](https://github.com/zulip/zulip/compare/a0d6b06a914e17d24f6890d2f23f90190d59bbf6..345a693ab9d4440c1b98d40b044ae72858cf2e30) and [here](https://github.com/zulip/zulip/compare/345a693ab9d4440c1b98d40b044ae72858cf2e30..4e09d5a908b4d7c7c75a93bfcb137d95f6528ff1)

shubham-padia commented:
@sahil839 pinging you for review since you were the reviewer of a previous PR

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -66,11 +67,34 @@ export class SettingsPanelMenu {
+        });
+        $("#admin-user-list .tab-container").append($(this.org_user_settings_toggler.get()));
+        if (section === "user-list-admin") {
+            this.org_user_settings_toggler.goto(user_settings_tab);
```
Is this needed? I guess call to `goto` in `activate_section_or_default` should be enough.

shubham-padia replied:
It was needed when activate_section_or_default was called before, changed the order of things.

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -148,15 +176,19 @@ export class SettingsPanelMenu {
         } else {
             this.$curr_li = $li_for_section;
         }
-
```
We avoid removing/adding lines not related to the changes.

sahil839 reviewed (commented):
Posted a couple of comments. Looks good otherwise.

shubham-padia commented:
Also replaced `append` with `html` in the latest pushed change, since there was a bug where clicking on `organization` tab will introduce more and more tab switchers because of the `append`

shubham-padia commented:
Bumping this for another review, thanks!

sahil839 commented on web/src/settings_panel_menu.js:
```diff
@@ -50,29 +51,63 @@ export class SettingsPanelMenu {
             // not to this click handler.
 
+            if (section === "user-list-admin") {
+                this.show_org_user_settings_toggler();
```
Can this just be called once in `activate_section_or_default` instead of having this here and in `show()` function as well?

shubham-padia replied:
Yes, fixed [here](https://github.com/zulip/zulip/compare/4e09d5a908b4d7c7c75a93bfcb137d95f6528ff1..881863a3c2531a7c7dd20c4e1fa68bdc28f38e1d)

sahil839 commented:
Looks good to me.

We still update the toggler container and register click handlers multiple times even when it is not required. I can think of two solutions - either using a condition like `if this.org_user_settings_toggler !== undefined` but in that case we would need to add some code to set it to undefined when closing the modal` or we just setup the toggler inside `settings_users.set_up` that would require some weird placing of variables to avoid import cycles. I am not sure whether it is fine as it or we want to use one of the two workarounds.

shubham-padia commented:
> closing the modalor we just setup the toggler insidesettings_users.set_up`

Yeah, this one might not be the most simple thing to figure out.

> using a condition like if this.org_user_settings_toggler !== undefined but in that case we would need to add some code to set it to undefined when closing the modal.

If I'm understanding this correctly, in this case, we will need to re-initialize the toggle component again and again, while in the current implementation, it is initialized only once.

We can add extra code to check whether the event listeners are set and whether the html is already there for the tab switcher, but I'm wondering whether it's worth the extra complexity of doing that. 

I've tried to optimise for similar (minor than this) things in the past, but have been pointed to https://zulip.readthedocs.io/en/latest/subsystems/performance.html to focus on making the code more readable. I'm still not sure where to draw the line between optimisation & micro-optimisation, so some input from your side on whether this is an optimisation worth doing or not would be helpful. If it's worth doing, I'm sure there will be a way to do it without running into import cycle problems.

sahil839 commented:
Yes, if we setup the toggler in `settings_users.set_up`. the toggler code will be run everytime the settings overlay is opened (including registering the event handlers), but not while changing the sections and navigating them while keeping it open.
And though the current implementation initializes the toggler code only once but updates the html elements and registers the click handlers everytime the user opens the "Users" section which is not necessary.

So, yeah that might affect performance, but I guess since the toggler code is only for the "Users" section it would be much better to keep it in `settings_users` file. But again we might want to figure out where to declare the variables to avoid import cycles and that might again affect readability.

Regarding the performance part maybe we can try seeing how much time it takes for the settings overlay to open in both the cases and see if we are fine with that. BTW we also have these tabs in "Bots" page and there the toggler code is called in `settings_bots.set_up`.

shubham-padia commented:
I've added another check [here](https://github.com/zulip/zulip/compare/881863a3c2531a7c7dd20c4e1fa68bdc28f38e1d..e6d5540f572bbf6b470ebb4ef455542c16340ae4) to make sure that we only replace the html and register the event handlers only when we need to. We still need that jquery check to run every now and then, but using `performance.now()` for before and after the check, the time varied between `0` and `0.1` ms, which I feel is an acceptable time.

alya commented:
We should show a lock next to the "Users" tab only if we'd show a lock next to all the tabs we are consolidating. We shouldn't show it if the user is allowed to send invites.

alya commented:
I think it'll look better to have more vertical space below the tabs. There's more space in the channels menu, and I think that's easier to visually parse.

![Screenshot 2024-06-07 at 10 01 28@2x](https://github.com/zulip/zulip/assets/2090066/0d4ac325-625e-4f7c-9b56-2cddb5dece4e)
![Screenshot 2024-06-07 at 10 01 12@2x](https://github.com/zulip/zulip/assets/2090066/e9defd0a-ecf8-483f-b47e-48e600ba9bdc)

alya commented:
Let's be sure to update the help center as an immediate follow-up to this, if we don't do it on this PR.

alya commented:
While we're here, let's rename the table Invites -> Invitations for consistency.

![Screenshot 2024-06-07 at 10 03 51@2x](https://github.com/zulip/zulip/assets/2090066/4e871b8a-21b6-4ea2-98f5-7cea3d55fd96)

alya commented:
I don't know if this should be a separate follow-up, but it's a bit confusing that when you deactivate/reactivate a user, they don't show up in the other table until you close and reopen the settings menu.

alya commented:
Not super important, but I noticed and filed #30353 while reviewing this PR.

timabbott commented on web/src/hashchange.js:
```diff
@@ -366,9 +366,9 @@ function do_hashchange_overlay(old_hash) {
-            settings_panel_menu.org_settings.activate_section_or_default(section);
+            settings_panel_menu.org_settings.set_current_tab(section);
         }
         settings_toggle.highlight_toggle(base);
```
Reading this, I kinda feel like `settings_toggle.highlight_toggle(base);` is misnamed; that's the function call here that's actually doing the navigation, yet it's very much sounding like it's just doing highlighting.

timabbott commented on web/src/components.ts:
```diff
@@ -115,10 +116,13 @@ export function toggle(opts: {
+        meta.$ind_tab.on("click", function () {
+            const idx = Number($(this).attr("data-tab-id"));
+            select_tab(idx);
+        });
```
This doesn't remove any existing event handlers; so we need to be careful using this API to avoid letting multiple get applied to the target element. Do we need an `off` removal first?

timabbott commented on web/styles/app_components.css:
```diff
@@ -316,6 +316,10 @@ input::placeholder {
 
+#admin-user-list .tab-switcher .ind-tab {
+    width: 110px;
+}
```
This seems like it belongs in `settings.css`, since it's styling for a specific element, not a reusable component.

shubham-padia replied:
Fixed!

sahil839 replied:
This can be removed now, right?

timabbott commented on web/src/hashchange.js:
```diff
@@ -72,6 +72,28 @@ function is_somebody_else_profile_open() {
+    const valid_user_settings_tab_values = new Set([
+        "active-users-admin",
+        "deactivated-users-admin",
+        "invites-list-admin",
```
These are user-facing names, since they appear in URLs, and we should probably pick something that actually looks nice. I see no reason for these to have extra tokens beyond just `active/deactivated/invites`.

We should probably discuss in #design, including whether `user-list-admin` should be renamed; I think `#organization/users/active` would be a much cleaner URL scheme for this section of the panel.

alya replied:
That sounds great as a scheme; I'd be fine with moving forward without a #design discussion.

timabbott replied:
OK but probably we should call it `/invitations` to match the new title.

shubham-padia replied:
I've renamed `user-list-admin` to `users` along with a redirect in ca22e48b8546ba89aa4f0f4545d2f32f6e8bc055, as a preparatory commit and changed the tab_values in the main commit.

timabbott commented on web/src/hashchange.js:
```diff
@@ -436,6 +462,9 @@ function do_hashchange_overlay(old_hash) {
         admin.build_page();
+        if (section === "user-list-admin" && settings_panel_menu.org_settings !== undefined) {
+            settings_panel_menu.org_settings.set_user_settings_tab(get_user_settings_tab(section));
+        }
```
This probably deserves a brief comment, like:

> The user settings panel has 3 sub-panels with different URLs; mark the correct one to open.

timabbott replied:
I feel like this is definitely a wart of the API that this PR defines; like why can't `admin.launch`, which is passed the full section including the subpanel, be the thing that's responsible for making sure we load the desired subpanel correctly?

To me, it feels like an abstraction violation that `hashchange.js` needs to be doing this setup.

shubham-padia replied:
Fixed !

timabbott commented:
lgtm aside from the issues noted above. After testing, think probably given the new UI structure, we do need to fix the live-update issue in this PR; the change in UX model of having these be a single panel means that it feels pretty broken to not have live-update working there.

alya commented:
I accidentally posted this comment from an earlier review on the wrong PR:

> Testing this PR, keyboard navigation with arrows between tabs doesn't seem to be working. (It works with the tabs in the channels menu.)

shubham-padia commented:
Live-update has been implemented in the second last commit.

Renamed `highlight_toggle` to `goto` in 93fbc3a0f0d7cad445822ca96b33a19c54138d99

> Testing this PR, keyboard navigation with arrows between tabs doesn't seem to be working. (It works with the tabs in the channels menu.)

Fixed in last commit, kept the commit seperate so it's easier to review.

> While we're here, let's rename the table Invites -> Invitations for consistency.

Fixed in 852f6feb2c20dcd5a2eb1f4358fe0322ac547b11

> Let's be sure to update the help center as an immediate follow-up to this, if we don't do it on this PR.

If the answer to both questions in this [CZO message](https://chat.zulip.org/#narrow/stream/19-documentation/topic/Document.20unified.20user.20settings.20panel/near/1824882) is no change required, then this PR updates the help center and fixes #28888

alya commented:
The live update and tab navigation works for me in manual testing!

alya commented:
Let's finish renaming invites -> invitations:

![Screenshot 2024-06-13 at 11 19 19](https://github.com/zulip/zulip/assets/2090066/40412e01-bcb4-48a5-8ca9-c781e55a49f4)

alya commented:
I replied about what we need to do for the help center on the CZO thread.

alya commented:
@sahil839 do you have any more feedback on this one?

sahil839 commented:
Posted one comment about removing unnecessary CSS.

I tested it and a couple of points -

- For live-updating, I deactivate a user from "Users" tab. It is still present in the "Users" tab which is correct. Then I go to "Deactivated" tab the just deactivated user is shown in the list. But then when I go back to "Users" tab, shouldn't the deactivated user be removed from the list. It was fine to keep the user there just after deactivation for allowing the user to reactivate, but once I move away from that tab, I think we can just remove that user from active users list.

- When I open the "Users" panel, navigating to different tabs using keyboard do not work till I click on one of the tabs once. I guess this problem is also present in the "Keyboard shortcuts" UI. So, might be fine to fix this later as a follow-up.

shubham-padia commented:
>When I open the "Users" panel, navigating to different tabs using keyboard do not work till I click on one of the tabs once. I guess this problem is also present in the "Keyboard shortcuts" UI. So, might be fine to fix this later as a follow-up.

This one is hard-ish to fix as part of this feature. The keyboard shortcuts in the settings panel are not using something like `hotkey.js` so it becomes a bit harder to manipulate. Ideally all the shortcuts for that panel should be going through one file where deciding the priority is easier but right now, when we are setting the key_handlers for each of these separately, it becomes harder to make complex operations.

> For live-updating, I deactivate a user from "Users" tab. It is still present in the "Users" tab which is correct. Then I go to "Deactivated" tab the just deactivated user is shown in the list. But then when I go back to "Users" tab, shouldn't the deactivated user be removed from the list. It was fine to keep the user there just after deactivation for allowing the user to reactivate, but once I move away from that tab, I think we can just remove that user from active users list.

This should be straightforward to implement, but do we need to discuss this on the `design` channel before we do this or not?

shubham-padia commented:
>I replied about what we need to do for the help center on the CZO thread.

In this case, #28888 is fixed in this PR, a review of the help center would be appreciated to check that.

> Let's finish renaming invites -> invitations:

Updated all user-facing strings on that page to use invitations including empty state messages.

sahil839 commented:
> This one is hard-ish to fix as part of this feature. The keyboard shortcuts in the settings panel are not using something like hotkey.js so it becomes a bit harder to manipulate. Ideally all the shortcuts for that panel should be going through one file where deciding the priority is easier but right now, when we are setting the key_handlers for each of these separately, it becomes harder to make complex operations. This would be better done as a follow up.

I agree. I also tried fixing the keyboard shortcuts overlay some time back but was not able to implement a clean working fix.

> This should be straightforward to implement, but do we need to discuss this on the frontend channel before we do this or not?

Would be good to get feedback first about what should be the expected behavior. Might be a good discussion for `design`,as we want to discuss about the behavior and not about implementation details.

alya commented:
Removing the user when you come back to the tab makes sense to me. I don't think we need a discussion.

alya commented:
The help center changes look correct to me; I haven't tried to thoroughly check for completeness.

alya commented:
@shubham-padia You can mark this PR for integration review once you've made the tweak noted above; we can leave fixing the keyboard nav for a follow-up.

shubham-padia commented:
> For live-updating, I deactivate a user from "Users" tab. It is still present in the "Users" tab which is correct. Then I go to "Deactivated" tab the just deactivated user is shown in the list. But then when I go back to "Users" tab, shouldn't the deactivated user be removed from the list. It was fine to keep the user there just after deactivation for allowing the user to reactivate, but once I move away from that tab, I think we can just remove that user from active users list.

Fixed [here](https://github.com/zulip/zulip/compare/fc7f70193e765e9ea0658d0cc73e8c0a21c7f691..697b9e7c11c635aa9e1cea4f0ae808e5320e76e3). In addition to the above requested behaviour, If you've deactivated a user on the `active` tab, the user will get greyed out. But if you `click` the `active` tab on top in the tab switcher, the greyed out user will disappear.

Also marked for `integration review`

timabbott commented:
Merged, thanks @shubham-padia and @sahil839 for all the work on this!
